### PR TITLE
Introduce value object mechanism

### PR DESCRIPTION
**Pull request by meaboutsoftware**
## 📋 Description
- This PR adds a reusable value object to the common package

## 📦 PR Includes

- [X] Feature added 🆕
- [ ] Bug fix 🐛
- [ ] Code refactor 🔄
- [ ] Documentation update 📚
- [ ] Tests added/updated 🧪
- [ ] Other: (describe) 

## 💡 Additional Notes

No additional notes

### CODE REVIEW

File: Chapter-4-applying-tactical-domain-driven-design/Fitnet.Common/Fitnet.Common.Core/ValueObject.cs [hunk 1]
```diff
@@ -0,0 +1,26 @@
+﻿namespace EvolutionaryArchitecture.Fitnet.Common.Core;
+
+public abstract class ValueObject
+{
+    protected abstract IEnumerable<object> GetEqualityComponents();
+
+    public override bool Equals(object? obj)
+    {
+        if (obj == null || obj.GetType() != GetType())
```
Comment (kamilbaczek) [inline Chapter-4-applying-tactical-domain-driven-design/Fitnet.Common/Fitnet.Common.Core/ValueObject.cs:9]: can we use "obj is null" ...? What  you think?

File: Chapter-4-applying-tactical-domain-driven-design/Fitnet.Common/Fitnet.Common.Core.UnitTests/ValueObjectTests.cs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    private class AnotherTypeFakeValueObject : ValueObject
+    {
+        protected override IEnumerable<object> GetEqualityComponents()
+        {
+            yield return DefaultIntProperty;
+            yield return DefaultStringProperty;
+        }
+    }
+}
+
```
Comment (kamilbaczek) [inline Chapter-4-applying-tactical-domain-driven-design/Fitnet.Common/Fitnet.Common.Core.UnitTests/ValueObjectTests.cs:87]: propably we can remove this empty line

File: Chapter-4-applying-tactical-domain-driven-design/Fitnet.Common/Fitnet.Common.Core/ValueObject.cs [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+            return false;
+        }
+
+        var other = (ValueObject)obj;
+
+        return GetEqualityComponents().SequenceEqual(other.GetEqualityComponents());
+    }
+
+    public override int GetHashCode() => GetEqualityComponents()
+            .Aggregate(0, (current, obj) => current ^ (obj?.GetHashCode() ?? 0));
```
Comment (kamilbaczek) [inline Chapter-4-applying-tactical-domain-driven-design/Fitnet.Common/Fitnet.Common.Core/ValueObject.cs:20]: public override int GetHashCode() =>
        GetAtomicValues().Aggregate(
            default(int),
            (hashcode, value) =>
                HashCode.Combine(hashcode, value.GetHashCode()));
                
                We could use potentially HashCode.Combine to make this caculation simpler. What you think?

File: Chapter-4-applying-tactical-domain-driven-design/Fitnet.Common/Fitnet.Common.Core/ValueObject.cs [hunk 4]
```diff
@@ -0,0 +1,26 @@
+﻿namespace EvolutionaryArchitecture.Fitnet.Common.Core;
+
+public abstract class ValueObject
```
Comment (kamilbaczek) [inline Chapter-4-applying-tactical-domain-driven-design/Fitnet.Common/Fitnet.Common.Core/ValueObject.cs:3]:
```suggestion
public abstract class ValueObject : IEquatable<ValueObject>
```

Can we implement interface IEquatable that will ensure that ValueObject behaves correctly and consistently with collections. 

What is your opinion about it?

Comment (kamilbaczek) [reply Chapter-4-applying-tactical-domain-driven-design/Fitnet.Common/Fitnet.Common.Core/ValueObject.cs:3]: We could as well add unit tests that test ensure value objects equalability works during comparision in collection for e.g in LINQ Where statment.

**Review by kamilbaczek — CHANGES_REQUESTED** [review]
please check comments

File: Chapter-4-applying-tactical-domain-driven-design/Fitnet.Common/Fitnet.Common.Core/ValueObject.cs (hunk 4 above)
Comment (meaboutsoftware) [reply Chapter-4-applying-tactical-domain-driven-design/Fitnet.Common/Fitnet.Common.Core/ValueObject.cs:3]: Great spot! I completely forgot about it :)

**Review by kamilbaczek — APPROVED** [review]

**Pull request merged**